### PR TITLE
[BitBucket GitHub] fixes and tests for BitBucket service integration

### PR DESCRIPTION
**Pull request by chris48s**
* Add test suite for BitBucket service integration
* Present BitBucket issues as a metric (for consistency with BitBucket PR endpoint and GitHub endpoints)
* Factor out a shared regex
* Fail cleanly if count is `undefined`

i.e: do this: invalid
not this: undefined_open

### CODE REVIEW

**Comment by chris48s** [comment]
I expect the GH failures are caused by #979 rather than stuff I've broken. If it makes it easier to review, I could back out 3611916 so only the BitBucket tests need to be run.

File: service-tests/helpers/validators.js [hunk 1]
```diff
@@ -37,6 +37,8 @@ const isStarRating = withRegex(/^(?=.{5}$)(\u2605{0,5}[\u00BC\u00BD\u00BE]?\u260
 // Required to be > 0, beacuse accepting zero masks many problems.
 const isMetric = withRegex(/^[1-9][0-9]*[kMGTPEZY]?$/);
 
+const isMetricOpenIssues = withRegex(/^[0-9]+[kMGTPEZY]? open$/);
```
Comment (paulmelnikow) [inline service-tests/helpers/validators.js]: Could we make sure this is nonzero, as with `isMetric`? We've had broken code which displayed 0 on a malformed response.

**Comment by paulmelnikow** [comment]
> I expect the GH failures are caused by #979 rather than stuff I've broken. If it makes it easier to review, I could back out 3611916 so only the BitBucket tests need to be run.

No need, I'll just run the Github tests locally before merging. They are green right now. ✅

File: service-tests/helpers/validators.js (hunk 1 above)
Comment (platan) [reply service-tests/helpers/validators.js]: Alternatively we could add a test with API mock which will allow expect specific value in result.

Comment (chris48s) [reply service-tests/helpers/validators.js]: It is legitimate for this to be zero if there are actually zero open issues though. In the GitHub tests (on master), you've got a mix of tests that are looking for `[0-9]+...` (e.g: https://github.com/badges/shields/blob/master/service-tests/github.js#L69-L74 ) and others looking for `[1-9][0-9]+...` (e.g: https://github.com/badges/shields/blob/master/service-tests/github.js#L80-L85 ).

Do you think we should extract both regexes and add tests for both cases, or should we always validate a non-zero value?

File: service-tests/bitbucket.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+t.create('issues-raw (valid)')
+  .get('/issues-raw/atlassian/python-bitbucket.json')
+  .expectJSONTypes(Joi.object().keys({
+    name: 'issues',
+    value: isMetric
+  }));
+
+t.create('issues-raw (invalid)')
+  .get('/issues-raw/atlassian/not-a-repo.json')
+  .expectJSON({ name: 'issues', value: 'invalid' });
```
Comment (platan) [inline service-tests/bitbucket.js]: How about displaying `repo not found` instead of `invalid`? Someone asked for repo which does not exist.

File: server.js [hunk 3]
```diff
@@ -4277,6 +4280,9 @@ cache(function(data, match, sendBadge, request) {
     try {
       var data = JSON.parse(buffer);
       var pullrequests = data.size;
+      if (typeof pullrequests === 'undefined') {
```
Comment (platan) [inline server.js]: In tests I can see that this code is used to handle case, when repo if not found (please correct me if I'm wrong). In this case API will return 404 status code (e.g. `https://bitbucket.org/api/2.0/repositories/atlassian/python-bitbucket/pullrequests_NOT_FOUND/?limit=0&state=OPEN`). How about checking status code before parsing response body?

**Comment by chris48s** [comment]
@platan - I can't think of another case where should be valid for count/size to be undefined with a `200 OK` response so I have changed the condition to check for `if (res.statusCode !== 200)` :+1:

File: service-tests/bitbucket.js (hunk 2 above)
Comment (chris48s) [reply service-tests/bitbucket.js]: 1. This block might potentially catch errors other than repo doesn't exist

2. This behaviour is consistent with the output of other endpoints. e.g:

* https://img.shields.io/cpan/v/not-a-package.svg
* https://img.shields.io/pypi/v/blablabla/not-a-package.svg
* https://img.shields.io/packagist/v/blablabla/not-a-package.svg

Comment (paulmelnikow) [reply service-tests/bitbucket.js]: Okay, though is there another way to detect the "not found" case? Agree with @platan that it would be better to detect it and return a helpful message. It lets the user know where the problem lies.

File: service-tests/helpers/validators.js (hunk 1 above)
Comment (paulmelnikow) [reply service-tests/helpers/validators.js]: True, though as long as we pick a project that is likely to stay around, there will always be open issues. I'm always fine with additional tests using mocks if the integrated tests don't seem satisfactory.

I think using `/^[1-9][0-9]*[kMGTPEZY]? open$/` would be fine for both open issues and open issues by label. Does that answer your question?

File: service-tests/bitbucket.js (hunk 2 above)
Comment (paulmelnikow) [reply service-tests/bitbucket.js]: Ah, didn't see this: https://github.com/badges/shields/pull/1315#discussion_r154759969

**Comment by chris48s** [comment]
updated based on latest review comments

**Comment by paulmelnikow** [comment]
Github service tests passing locally 👍 

Thanks!

File: server.js [hunk 4]
```diff
@@ -4246,11 +4246,18 @@ cache(function(data, match, sendBadge, request) {
     try {
       var data = JSON.parse(buffer);
       var issues = data.count;
-      badgeData.text[1] = issues + (isRaw? '': ' open');
+      if (res.statusCode !== 200) {
```
Comment (platan) [inline server.js]: It's already merged, but I think next time we could check status code before parsing response body.
